Improve demo setup docs

bedengler opened the pull request:
The atk4/ui demo now runs also if you position it outside the atk4 repo and rename the "demos" directory to something else. This should add a layer of positive user experience, if someone tries that out and it works flawlessly.

In addition I have added documentation for & link to the atk4/ui demos.

Commit:
atk4 demo outside repo & documentation

The atk4/ui demo now runs also if you put it outside the original atk4 location. This adds a layer of positive user experience, if someone tries it out and it works.

In addition I have added documentation for & link to the atk4 demos.

Commit:
Unit tests didn't run - testing to change back

just testing to change some things for unit tests

mvorisek commented on docs/overview.rst:
```diff
@@ -451,3 +451,51 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
 hosting that runs PHP apps.
 
+
+Online-Demo showing many functions
```
this is ok if not written somewhere else alredy

bedengler replied:
I didn't find it anywhere. Even if it's anywhere else, it should be placed in the overview (which is the first place when  someone is new to atk4 and opens the documentation).
And it should be removed from "anywhere else", if someone could find it.

mvorisek commented on docs/overview.rst:
```diff
@@ -451,3 +451,51 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
+https://agile-ui.readthedocs.io/en/latest/index.html
+
+
+Demo contained in the atk4 repository
```
text below is quite specifically worded...

there should be:
- advise demos need clone/fork
- how sqlite database can be created: `php demos/_demo-data/create-sqlite-db.php`

bedengler replied:
According to @georgehristov they are adviced to don't move the original repos and use them as they are.
What do we advice for consistency?  

1. clone just the "demos" directory or
2. clone the entire atk4/ui directory or
3. don't clone and use the files in /vendors/atk4/ui...

???

bedengler replied:
@mvorisek, I added SQLITE but didn't commit. I'm waiting for feedback for the cloning issue...

mvorisek replied:
fork whole atk4/ui + composer install + init sqlite db, then make whole atk4/ui accessible by webserver

bedengler replied:
What about moving just the "demos" directory in a subdirectory called "atk4/ui" on the webserver that's publicly available instead of the whole atk4/ui directory? 
Advantage: when updating atk4/ui, you have the demos separated and they aren't overwritten.

mvorisek replied:
impossible - you need to fork the whole ui

for separate vendor files, you need to understand composer more deeply and I think this should be not documented as too advanced for most users...

bedengler replied:
adapted, please check

Commit:
Adaptions for more clarity

Changes and enhancements according to 4de5b296db4ec42e38e37d93407601b50a87243c and db0da0ce418047ced7e1eed25a276ddd042200ad.

Additionally I added some examples for CRUD for more convenience. Finding those took me hours...

mvorisek commented on demos/init-app.php:
```diff
@@ -51,8 +51,13 @@
+// Current sub directory of webserver - makes it independent where to run the demo from
+// basically we want to run it just from within the ui repo, but this adds a layer of
+// better user experience, when a user tries to copy it elsewhere and it works
+$currDir = substr(str_replace($_SERVER['DOCUMENT_ROOT'], "", dirname(__FILE__)), 1);
```
I belive should be not supported, we should actually not distribute tests/demos in release at all for security reasons.

bedengler replied:
I'm seeing it from the perspective as a user (I for myself joined atk4 a year ago). 
The learning curve is ridiculous steep as there's a big lack of documentation. Reading everything out of the code without having examples makes everything even more time consuming. Which is unnecessary imo, if we just 
1) offer more examples
2) write documentations
3) make it as easy as possible to get results (raise usability)

Currently you have to try out a lot. 
If you just spend hours or days of getting a demo running, users jump off again before they even were able to see a welcome screen...

Therefore I'd be interested in opinions of other team members first

mvorisek replied:
please post usecase (step by step) that needs this change

romaninsh replied:
I think this is a positive change and it does not break anything. Sound reasoning on moving things outside.

mvorisek replied:
> 
> 
> please post usecase (step by step) that needs this change

as written, I need usecase that needs this change, otherwise `str_replace($_SERVER['DOCUMENT_ROOT'], "", dirname(__FILE__)` seems super hacky to me and it will not work on even slighly badly configured servers!

mvorisek commented on docs/overview.rst:
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
+"atk4/ui/demos/_demo-data/".
+Call it as follows (you may have to adapt the paths to fit your webserver structure)::
+
+    php atk4/ui/demos/_demo-data/create-sqlite-db.php;  // when using the CLI approach
```
we should allow to create from CLI only

romaninsh replied:
it might be more confusing to offer alternative implementation, but if your opinion as someone new to ATK that both instruction should be provided -that's OK.. 

Just consider file permission implications. 12-factor app requires applications to be runnable in read-only containers.

Same can be solved with docker-compose and mysql container, btw.

mvorisek commented on docs/overview.rst:
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
+Additional tips
+---------------
+
+ - If you are using Eclipse, import the previously cloned "atk4" directory into Eclipse (then it finds all sub-projects)
```
no specific doc for Eclipse

romaninsh replied:
in my case directory is always called "ui", not atk4.

Commit:
Adaptions for testing

Tests failed due to existence of a variable that hasn't been unset

Commit:
Merge remote-tracking branch 'upstream/develop' into develop
Merged because I merged berfore loading upstream

mvorisek commented:
@bedengler revert `demos/init-app.php`, then we @ibelar or @georgehristov can review or we can close completely

romaninsh commented on docs/overview.rst:
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
+repository which also could be locally executed.
+This folder is basically intended to run under the UI package only.
+That means the demo may not work properly, if you copy the "demos" directory out of this location
+(to run it from a different location) or if the "vendors" directory isn't accessible through the
```
I thought the above fix addresses that? 

Could you rephrase this text in the form of step-by-step instruction on how to actually run demo elsewhere?

romaninsh commented:
@bedengler - thanks so much for not only identifying documentation issue, but also be willing to improve the situation. Framework relies on contributors and we need more people focused specifically on the docs, so if you are serious about this - you would be able to refine documentation gradually.

Overall - your instruction seem very "specific" to your personal use case, can require a bit more work, but having this much is better than having nothing at all, so well done!

mvorisek reviewed (changes requested):
need changes

mvorisek committed:
Merge branch 'develop' into bedeng_develop

mvorisek committed:
revert all

mvorisek committed:
cherrypick useful

mvorisek commented:
@bedengler Please proof read the changes, remove what is not needed/up-to-date and we can merge then.

abbadon1334 reviewed (approved):
LGTM
